core/src/main/java/org/apache/spark/shuffle/sort/ShuffleExternalSorter.java

```diff
@@ -321,9 +320,10 @@ private void growPointerArrayIfNecessary() throws IOException {
     assert(inMemSorter != null);
     if (!inMemSorter.hasSpaceForAnotherRecord()) {
       long used = inMemSorter.getMemoryUsage();
-      long needed = used + inMemSorter.getMemoryToExpand();
+      MemoryBlock page;
       try {
-        acquireMemory(needed);  // could trigger spilling
+        // could trigger spilling
+        page = taskMemoryManager.allocatePage(used * 2, this);
```

Comment:
Implicit here is the fact that the in memory sorter's only source of memory usage is the pointer array itself. That's fine, though.